### PR TITLE
Add Numeric literals in daml-lf-ast and daml-lf-proto.

### PR DESCRIPTION
**Pull request by associahedron**
This PR adds the Numeric literal in daml-lf-ast, encoding/decoding for it in daml-lf-proto, and type checking, etc as part of #2289.

The (soon-to-be-deprecated) Decimal literal was represented with a fixed decimal point number (`Fixed E10`). This doesn't work for Numerics since the whole point is to have a parametrized decimal point number. Numeric literals in DAML-LF carry the scale implicitly (as the number of digits after the decimal point), so we need to both store the value and keep track of the scale.

I tried to find an existing decimal library that would do this for us, and the closest I found was `Data.Decimal`. The module `DA.Daml.LF.Ast.Numeric` implements a `Numeric` type for DAML-LF Numeric literals, based on `Data.Decimal`, with a few adjustments. I added a number of tests to make sure we're handling read/show correctly for Numeric literals.

### CODE REVIEW

File: compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Numeric.hs [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+import Control.Monad
+import Data.Data
+import Data.Decimal
+import Data.Maybe
+import GHC.Generics (Generic)
+import Numeric.Natural
+
+-- | Numeric literal. This must encode both the mantissa (up to 38 digits) and
+-- the scale (0-37), the latter controlling how many digits appear after the
+-- decimal point. Furthermore, when reading or writing a numeric instance,
```
Comment (cocreature) [inline compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Numeric.hs]: `instance` isn’t a typeclass instance right? I guess you mean a value of type `Numeric`?

File: compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Numeric.hs [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+-- we need to show every significant digit after the decimal point, in order
+-- to preserve the scale. For scale 0, the decimal point needs to be shown
+-- without a digit.
+--
+-- Internally we use Data.Decimal to represent these because it has the ability
+-- to encode every Numeric alongside its scale, and it mostly does what we want
+-- with Show and Read, with a few adjustments:
+--
+-- * we perform bounds checks with smart constructor 'numeric'
+-- * for scale 0, we have to handle the decimal point in the Show
```
Comment (cocreature) [inline compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Numeric.hs]: Thanks for the great explanations!

File: compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Numeric.hs [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+-- (inclusive).
+numericMantissa :: Numeric -> Integer
+numericMantissa = decimalMantissa . numericDecimal
+
+instance Show Numeric where
+    showsPrec p n
+        | numericScale n == 0 = shows (numericDecimal n) . ("." ++)
+        | otherwise = showsPrec p (numericDecimal n)
+
+instance Read Numeric where
```
Comment (cocreature) [inline compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Numeric.hs]: Maybe this is a silly question but does DAML compilation rely on the `Show` and `Read` instances or is this just for tests/experimentation?

File: compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Numeric.hs [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+                case xs of
+                    '.':ys -> Just (n, ys)
+                    _ -> Nothing
+
+        numericValid :: Numeric -> Bool
+        numericValid n =
+            numericScale n <= numericMaxScale
+            && numericMantissa n <= numericMaxMantissa
+
+instance Data Numeric where
```
Comment (cocreature) [inline compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Numeric.hs]: Might be reasonable to propose the instance upstream.

File: compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Numeric.hs (hunk 3 above)
Comment (associahedron) [reply compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Numeric.hs]: They're used in the encoding and decoding of decimal/numeric literals in daml-lf-proto, and in the Pretty instance. Not sure how critical, but I don't want to be introducing weird behavior there :-)

Comment (cocreature) [reply compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Numeric.hs]: Ah I see, totally agree that we should get this right.

File: compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Numeric.hs (hunk 1 above)
Comment (associahedron) [reply compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Numeric.hs]: Nice catch! :)

File: compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Numeric.hs (hunk 4 above)
Comment (associahedron) [reply compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Numeric.hs]: Yes, that's a good idea :)

File: compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Serializability.hs [hunk 5]
```diff
@@ -54,13 +55,13 @@ serializabilityConditionsType world0 _version mbModNameTpls vars = go
...
           | otherwise -> Left (URNumericOutOfRange n)
       TNumeric _ -> Left URNumericNotFixed
           -- We statically enforce bounds check for Numeric type,
-          -- requiring 0 <= n <= 38 for the argument to Numeric.
-          -- If the argument isn't given explicitly, we can't
-          -- guarantee serializability.
+          -- requiring 0 <= n <= 'numericMaxScale' for the argument
+          -- to Numeric. If the argument isn't given explicitly, we
+          -- can't guarantee serializability.
       TNat _ -> noConditions
```
Comment (hurryabit) [inline compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Serializability.hs]: `TNat` should not be serializable at all. Most client languages can't express this. The only place where it's allowed in serializable types is immediately under `TNumeric`. I must have missed this in the previous PR.

Comment (associahedron) [reply compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Serializability.hs]: That makes sense. I'll update it, thanks!

**Pull request merged**